### PR TITLE
API for accessing db instances outside handler

### PR DESCRIPTION
**Pull request by gigioSouza**
API for accessing db instances outside handler, as mentioned in #18 

Usage:
```
const instances = require('hapi-sequelizejs').instances;

const dbs = instances.dbs; 
// returns dbs instances as key value object { [db.name]: db.instance }
dbs.myDb;

const myDb = instances.getDb('myDb');
// returns the given name db instance
// throws Error if not found

const myFirstRegisteredDb = instances.getDb();
// returns the first registered db instance
// throws Error there isn't any db instance

const myDbModels = instances.getModels('myDb');
// returns the models from the given name db instance
// throws Error if db instance not found

const myFirstRegisteredDbModels = instances.getModels();
// returns the models from the first registered db instance
// throws Error there isn't any db instance

const myDbMyModel = instances.getModel('myDb', 'myModel');
// returns the model from the given name from the given db's name instance
// throws Error if db instance not found
// throws Error if model not found in db instance

const myFirstRegisteredDbMyModel = instances.getModel('myModel');
// returns the model from the given name from the first registered db instance
// throws Error there isn't any db instance
// throws Error if model not found in db instance
```

### CODE REVIEW

**Comment by gigioSouza** [comment]
Sorry for the multiple commits, I just realized I could get the coverage data while in development without commiting, I'll create tests for the missing cases and reopen the PR.

**Comment by Nargonath** [comment]
After just a quick scan, the `package.json` is not committed hence you should consider removing your `yarn.lock` and `yarn-error.log` from your commits and also add these files to the `.gitignore`. They are adding noise to the PR which is not required.

**Comment by Nargonath** [comment]
You did add the `yarn.lock` file into the gitignore file so further modification will not be tracked but the version you committed is still there. IIRC you need to do `git rm --cached yourFile` to delete it from git.

**Comment by gigioSouza** [comment]
@Nargonath Didn't noticed I kept the `yarn.lock` before inserting it to the `.gitignore` and didn't know that command either, thanks!

**Comment by Nargonath** [comment]
@gigioSouza No problem, you're welcome. :+1:

**Comment by valtlfelipe** [comment]
I have not forgot about this. As soon as possible I will have a look at this

**Comment by gigioSouza** [comment]
@valtlfelipe thanks, I'm anxious for my "first" open source contribution

**Comment by Samueljoli** [comment]
@gigioSouza @valtlfelipe Hey folks! I was wondering if there's been any updates on this?

**Comment by gigioSouza** [comment]
@Samueljoli I'm just waiting

**Comment by Nargonath** [comment]
I've been thinking about this lately and I wonder if that would not be better to give access to the `db` element through `server.app`. This is worth reading: https://hapipal.com/best-practices/server-plugin-separation

File: lib/DB.js [hunk 1]
```diff
@@ -5,11 +5,11 @@ class DB {
         this.sequelize = sequelize;
         this.models = models;
     }
-
+
```
Comment (Nargonath) [inline lib/DB.js]: This should not happen because `prettier` should remove useless space characters. Even if you don't have it setup in your IDE, the `precommit` hook should not let that pass through. It results in having a diff on file for no reason. Is your local project dependencies up-to-date? Did you see `precommit` hook running when you submit your PR? Asking because there might be an issue with that.

File: lib/index.js [hunk 2]
```diff
@@ -25,92 +6,95 @@ const Schema = require('./schema');
 const Models = require('./models');
 const DB = require('./DB');
 const Pkg = require('../package.json');
+const instances = require('./instances');
 
-// Module globals
-const internals = {};
+module.exports = {
```
Comment (Nargonath) [inline lib/index.js]: Refactoring the codebase to make it more readable is a nice thought and is welcomed but IMO it would be better done as a separate PR because otherwise it just adds noises to your PR and make it harder to review the real changes introduced by the new feature. AFAICT the refactoring is not needed for your feature to work, is it?

File: lib/models.js [hunk 3]
```diff
@@ -33,20 +49,15 @@ async function load(files, fn) {
 }
 
 async function applyRelations(models) {
-    if (!models || typeof models !== 'object')
-        throw new Error(`Can't apply relationships on invalid models object`);
-
+    if (typeof models !== 'object') {
```
Comment (Nargonath) [inline lib/models.js]: Why did you remove the `!models` tests? AFAICT the test is needed because if you get `models = null` this test will pass and `Object.keys()` will crash because it is given `null` as an argument. We need to test that `models` is an object but also not `null` because `typeof null` is `'object'`.

Comment (gigioSouza) [reply lib/models.js]: First I removed it for best test coverage, `models` will never be `null` in this code.
As you can see [here](https://github.com/valtlfelipe/hapi-sequelizejs/pull/19/files/a67b1552c019ef7ac3af68d4d17938609b5955f9#diff-6d186b954a58d5bb740f73d84fe39073R65), the `null check` is made before the call.

File: lib/DB.js (hunk 1 above)
Comment (gigioSouza) [reply lib/DB.js]: My dependencies are up-to-date, I didn't noticed if the `precommit` hook run ok, sorry.

File: lib/index.js (hunk 2 above)
Comment (gigioSouza) [reply lib/index.js]: Yes, it works without the refactoring. I refactored it to improve the readability and for understand the library to insert my proposal. I didn't think about doing this into another PR. I can send it into another PR if you prefer.

**Comment by gigioSouza** [comment]
I'll open a new PR with the refactory I did into the project, as I promised in this [comment](https://github.com/valtlfelipe/hapi-sequelizejs/pull/19#discussion_r186084395). As I'm new to node and hapijs, I have no position about the [comment](https://github.com/valtlfelipe/hapi-sequelizejs/pull/19#issuecomment-386330457).

**Comment by valtlfelipe** [comment]
@gigioSouza I will review the other PR, and merge it to master, then, please open a new PR with your implementation. This PR is already to messy.

On the new PR, please update it with master, so, it will have the refactoring, and the changes will only be your new code.

**Comment by gigioSouza** [comment]
@valtlfelipe We can still keep this PR, I just need to merge both branches. After you merge the refactory PR, only the implementation code will be kept in this PR.

File: README.md [hunk 4]
```diff
@@ -88,7 +88,9 @@ server.plugins['hapi-sequelizejs'][opts.name] = new DB(opts.sequelize, models);
 
 ### API
 
-#### `getDb(name)`
+#### Using `request` object
+
+#####`getDb(name)`
```
Comment (valtlfelipe) [inline README.md]: This broke the markdown, please add a space after the `#####`

**Comment by gigioSouza** [comment]
@valtlfelipe Merges done, now you may review

File: lib/instances.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+function getDb(dbName) {
+    let instance;
+    if (dbName) {
+        instance = findDb(dbName);
+    } else {
+        instance = instances[0];
+    }
+    
+    if (!instance) {
+        throw new Error('Database not found');
```
Comment (valtlfelipe) [inline lib/instances.js]: Please use the same message as for the server implementation `hapi-sequelizejs cannot find the ${name} database instance`

File: lib/instances.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        instance = getDb(dbName);
+    } else {
+        modelName = dbName;
+        instance = getDb();
+    }
+    
+    model = instance.models[modelName];
+    
+    if (!model) {
+        throw new Error('Database model not found');
```
Comment (valtlfelipe) [inline lib/instances.js]: please follow the same pattern as other messages

File: README.md (hunk 4 above)
Comment (gigioSouza) [reply README.md]: Fixed in 898c7885954ff24f67da4c35eb157e040fe37d3d

File: lib/instances.js (hunk 5 above)
Comment (gigioSouza) [reply lib/instances.js]: Fixed in 898c7885954ff24f67da4c35eb157e040fe37d3d

**Comment by gigioSouza** [comment]
@valtlfelipe I did changed the plugin to use the `instances`, [take a look](https://github.com/valtlfelipe/hapi-sequelizejs/pull/19/files#diff-6d186b954a58d5bb740f73d84fe39073R91)

**Comment by gigioSouza** [comment]
@valtlfelipe Now I realized, I misunderstood what you said [here](https://github.com/valtlfelipe/hapi-sequelizejs/pull/19#issuecomment-389543812). 
Does your idea is have access to the `instances` in the `request` object, right?

**Comment by valtlfelipe** [comment]
@gigioSouza No, I meant what you implemented, that the plugin makes use of the instances, so we have no duplicated code

**Comment by gigioSouza** [comment]
@valtlfelipe Well, take a look in 589b37b19323f649a4e8c3916ce674968745eab7
I added the `getDb`, `getModel` and `getModels` to the `request` object.

File: README.md [hunk 7]
```diff
@@ -102,14 +104,82 @@ handler(request, reply) {
...
+
+```
+const instances = require('hapi-sequelizejs').instances;
+
+const myDb = instances.getDb('myDb');
+
+const firstRegisteredDb = instances.getDb();
+```
+
+> If there isn't a db instance for the given name or no registered db instance, an Error is thrown: `Database not found`.
```
Comment (valtlfelipe) [inline README.md]: These messages have changed, please update the README

**Comment by geun** [comment]
+1

**Comment by gigioSouza** [comment]
@valtlfelipe sorry the delay to update the README. Please check if it's ok now.

**Comment by valtlfelipe** [comment]
Released 1.2.0 with this. Thanks for the effort @gigioSouza !! 😄
